### PR TITLE
[OSDEV-1116] Implement contribution record page

### PR DESCRIPTION
**Pull request by Innavin369**
[OSDEV-1116](https://opensupplyhub.atlassian.net/browse/OSDEV-1116) - A new Contribution Record Page has been developed to enable quick identification and moderation of contributions. This page includes two main sections: Moderation Event Data and Potential Matches, along with a set of buttons designed to facilitate the moderation process.

Screenshot 2024-11-05 at 14 31 24

[OSDEV-1116]: https://opensupplyhub.atlassian.net/browse/OSDEV-1116?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/react/src/util/styles.js [hunk 1]
```diff
@@ -654,6 +654,78 @@ export const makeDownloadExcelButtonStyles = theme =>
...
+            display: 'flex',
+            flexDirection: 'row',
+            justifyContent: 'space-between',
+        },
+        listItemTextStyle: {
+            padding: 0,
+        },
+        dividerStyle: {
+            height: '3px',
+            backgroundColor: 'rgb(0 0 0 / 55%)',
```
Comment (mazursasha1990) [inline src/react/src/util/styles.js:694]: Please add this color to the COLOURS.js.

File: src/react/src/util/styles.js
Comment (mazursasha1990) [inline src/react/src/util/styles.js:1]: Please check the colors that you have added, maybe some of them could be replaced by the existing ones.

**Review by mazursasha1990 — CHANGES_REQUESTED** [review]

**Review by mazursasha1990 — CHANGES_REQUESTED** [review]
Please, check the Browser console on the Contribution Record Page for errors.

File: src/react/src/__tests__/components/ContributionRecord.test.js
[the remark this answers is not in the document]
Comment (roman-stolar) [reply src/react/src/__tests__/components/ContributionRecord.test.js:1]: +1

File: src/react/src/__tests__/components/ContributionRecord.test.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    const rejectButton = getByRole('button', { name: /Reject Contribution/i });
+    const claimButton = getByRole('button', { name: /Go to Claim/i });
+
+    expect(createButton).toBeDisabled();
+    expect(rejectButton).toBeDisabled();
+    expect(claimButton).toBeDisabled();
+  });
+
+  test('should enable button when fetching is false', () => {
+    const {getByRole} = renderComponent({fetching: false,moderationID :'2'});
```
[the remark this answers is not in the document]
Comment (roman-stolar) [reply src/react/src/__tests__/components/ContributionRecord.test.js:58]: It should be 'uuid' format

File: src/react/src/components/Dashboard/DashboardContributionRecord.jsx [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+                                                <Button
+                                                    color="secondary"
+                                                    variant="contained"
+                                                    className={
+                                                        classes.confirmButtonStyles
+                                                    }
+                                                    onClick={() => {}}
+                                                >
+                                                    Confirm
+                                                </Button>
```
[the remark this answers is not in the document]
Comment (roman-stolar) [reply src/react/src/components/Dashboard/DashboardContributionRecord.jsx:149]: If the implementation didn't exist we can create some method/handler for the future (optional)

File: src/react/src/util/styles.js (hunk 1 above)
Comment (Innavin369) [reply src/react/src/util/styles.js:694]: replaced with COLOURS.DARK_GREY

File: src/react/src/util/styles.js
Comment (Innavin369) [reply src/react/src/util/styles.js:1]: done

File: src/react/src/util/styles.js [hunk 4]
```diff
@@ -654,6 +654,78 @@ export const makeDownloadExcelButtonStyles = theme =>
...
+            maxHeight: '300px',
+            overflow: 'scroll',
+        },
+        potentialMatchesBlock: {
+            width: '100%',
+            marginBottom: '25px',
+        },
+        potentialMatchesInternalBlock: {
+            width: '100%',
+            overflow: 'scroll',
```
[the remark this answers is not in the document]
Comment (Innavin369) [reply src/react/src/util/styles.js:710]: changed

File: src/react/src/util/styles.js [hunk 5]
```diff
@@ -654,6 +654,78 @@ export const makeDownloadExcelButtonStyles = theme =>
...
+            height: '3px',
+            backgroundColor: 'rgb(0 0 0 / 55%)',
+        },
+        innerDividerStyle: {
+            backgroundColor: 'rgb(0 0 0 / 55%)',
+        },
+        prettyPrint: {
+            width: '100%',
+            maxHeight: '300px',
+            overflow: 'scroll',
```
[the remark this answers is not in the document]
Comment (Innavin369) [reply src/react/src/util/styles.js:702]: changed

File: src/react/src/components/Dashboard/DashboardContributionRecord.jsx [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+            <div className={classes.buttonsContainerStyles}>
+                <Button
+                    color="primary"
+                    variant="contained"
+                    onClick={() => {}}
+                    className={classes.buttonStyles}
+                    disabled={fetching}
+                >
+                    Create New Location
+                </Button>{' '}
```
[the remark this answers is not in the document]
Comment (Innavin369) [reply src/react/src/components/Dashboard/DashboardContributionRecord.jsx:184]: removed

File: src/react/src/components/Dashboard/DashboardContributionRecord.jsx [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+                <Button
+                    color="secondary"
+                    variant="contained"
+                    onClick={() => {}}
+                    className={classes.buttonStyles}
+                    disabled={fetching}
+                >
+                    Reject Contribution
+                </Button>{' '}
```
[the remark this answers is not in the document]
Comment (Innavin369) [reply src/react/src/components/Dashboard/DashboardContributionRecord.jsx:193]: removed

File: src/react/src/components/Dashboard/DashboardContributionRecord.jsx [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+                        </ShowOnly>
+                        <ShowOnly
+                            key="section"
+                            when={
+                                potentialMatches && potentialMatches.length > 0
+                            }
+                        >
+                            {potentialMatches &&
+                                potentialMatches.map(
+                                    (potentialMatch, index) => (
```
[the remark this answers is not in the document]
Comment (Innavin369) [reply src/react/src/components/Dashboard/DashboardContributionRecord.jsx:110]: done

File: src/react/src/components/Dashboard/DashboardContributionRecord.jsx [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+                                        <>
+                                            {' '}
```
[the remark this answers is not in the document]
Comment (Innavin369) [reply src/react/src/components/Dashboard/DashboardContributionRecord.jsx:112]: removed

File: src/react/src/components/Dashboard/DashboardContributionRecord.jsx [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+                                </Typography>
+                            </div>
+                        </ShowOnly>
+                        <ShowOnly
+                            key="section"
+                            when={
+                                potentialMatches && potentialMatches.length > 0
+                            }
+                        >
+                            {potentialMatches &&
```
[the remark this answers is not in the document]
Comment (Innavin369) [reply src/react/src/components/Dashboard/DashboardContributionRecord.jsx:108]: right, removed

File: src/react/src/components/Dashboard/DashboardContributionRecord.jsx [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+                                                <Divider
+                                                    className={
+                                                        classes.innerDividerStyle
+                                                    }
+                                                    component="li"
+                                                />
+                                            </ShowOnly>
+                                        </>
+                                    ),
+                                )}
```
[the remark this answers is not in the document]
Comment (roman-stolar) [reply src/react/src/components/Dashboard/DashboardContributionRecord.jsx:169]: +1

**Comment by roman-stolar** [comment]
The request changes looks valid please resolve all necessary conversation than I also approved the PR.

File: src/react/src/components/Dashboard/DashboardContributionRecord.jsx [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+    fetchEvent: func.isRequired,
+    fetchMatches: func.isRequired,
+    classes: object.isRequired,
+    error: string,
+    fetchPotentialMatchError: string,
+};
+
+const mapStateToProps = ({
+    dashboardContributionRecord: {
+        moderationEvent: { event, fetching, error },
```
[the remark this answers is not in the document]
Comment (Innavin369) [reply src/react/src/components/Dashboard/DashboardContributionRecord.jsx:228]: renamed

File: src/react/src/util/propTypes.js [hunk 13]
```diff
@@ -511,3 +511,54 @@ export const moderationEventsPropType = arrayOf(
...
+        minimum_order_quantity: string,
+        average_lead_time: string,
+        percent_female_workers: 0,
+        affiliations: arrayOf(string),
+        certifications_standards_regulations: arrayOf(string),
+        historical_os_id: arrayOf(string),
+        country: productionLocationCountryPropType,
+        claim_status: string,
+    }),
+);
```
[the remark this answers is not in the document]
Comment (Innavin369) [reply src/react/src/util/propTypes.js:564]: fixed

File: src/react/src/reducers/DashboardContributionRecordReducer.js [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+            update(state, {
+                potentialMatches: {
+                    fetching: { $set: initialState.potentialMatches.fetching },
+                    error: { $set: initialState.potentialMatches.error },
+                    potentialMatches: { $set: payload },
+                },
+            }),
+    },
+    initialState,
+);
```
[the remark this answers is not in the document]
Comment (Innavin369) [reply src/react/src/reducers/DashboardContributionRecordReducer.js]: added

File: src/react/src/util/styles.js [hunk 15]
```diff
@@ -654,6 +654,78 @@ export const makeDownloadExcelButtonStyles = theme =>
...
+            margin: '20px',
+            padding: '20px',
+            width: '100%',
+            fontWeight: 'bold',
+        }),
+        container: {
+            marginBottom: '25px',
+            width: '100%',
+            padding: '20px',
+        },
```
[the remark this answers is not in the document]
Comment (Innavin369) [reply src/react/src/util/styles.js:680]: done

File: src/react/src/__tests__/components/ContributionRecord.test.js
[the remark this answers is not in the document]
Comment (Innavin369) [reply src/react/src/__tests__/components/ContributionRecord.test.js:1]: done

File: src/react/src/__tests__/components/ContributionRecord.test.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (Innavin369) [reply src/react/src/__tests__/components/ContributionRecord.test.js]: removed

File: src/react/src/components/Dashboard.jsx
[the remark this answers is not in the document]
Comment (Innavin369) [reply src/react/src/components/Dashboard.jsx:1]: done

File: src/react/src/__tests__/components/DashboardModerationQueueListTable.test.js
[the remark this answers is not in the document]
Comment (Innavin369) [reply src/react/src/__tests__/components/DashboardModerationQueueListTable.test.js:1]: reverted

File: src/react/src/components/Dashboard/DashboardContributionRecord.jsx [hunk 16]
```diff
@@ -0,0 +1,261 @@
...
+    recordList: {
+        py: 0,
+        width: '100%',
+        maxHeight: 360,
+        borderRadius: 2,
+        border: '3px solid',
+        borderColor: 'divider',
+        backgroundColor: 'background.paper',
+    },
+};
```
[the remark this answers is not in the document]
Comment (Innavin369) [reply src/react/src/components/Dashboard/DashboardContributionRecord.jsx:35]: doesn't work, inline style in this case works fine

File: src/react/src/util/styles.js [hunk 17]
```diff
@@ -654,6 +654,78 @@ export const makeDownloadExcelButtonStyles = theme =>
...
+            overflow: 'auto',
+            maxHeight: '450px',
+        }),
+        buttonsContainerStyles: Object.freeze({
+            display: 'flex',
+            flexDirection: 'row',
+            width: '100%',
+        }),
+        confirmButtonStyles: Object.freeze({
+            backgroundColor: '#b1dbb1',
```
Comment (mazursasha1990) [inline src/react/src/util/styles.js:719]: Please use `COLOURS.MINT_GREEN` here.

File: src/react/src/util/styles.js [hunk 18]
```diff
@@ -654,6 +654,78 @@ export const makeDownloadExcelButtonStyles = theme =>
...
+            padding: '20px',
+            '&:hover': { backgroundColor: '#799679' },
```
Comment (mazursasha1990) [inline src/react/src/util/styles.js:721]: Please add '#799679' to the COLOURS.js and name It some similar to MINT_GREEN.

File: src/react/src/util/styles.js [hunk 19]
```diff
@@ -654,6 +654,78 @@ export const makeDownloadExcelButtonStyles = theme =>
...
+        }),
+        claimButtonStyles: Object.freeze({
+            backgroundColor: theme.palette.action.main,
+            '&:hover': { backgroundColor: 'rgb(175 142 42)' },
```
Comment (mazursasha1990) [inline src/react/src/util/styles.js:725]: Please use `theme.palette.action.dark` here.

File: src/react/src/util/styles.js [hunk 20]
```diff
@@ -654,6 +654,78 @@ export const makeDownloadExcelButtonStyles = theme =>
...
+            display: 'block',
+            margin: 'auto',
+        }),
+        emptyBlockStyles: Object.freeze({
+            height: '100px',
+            display: 'flex',
+            justifyContent: 'center',
+            alignItems: 'center',
+        }),
+        emptyTextStyle: Object.freeze({ color: 'rgb(142 138 138 / 87%)' }),
```
Comment (mazursasha1990) [inline src/react/src/util/styles.js:669]: Please use `theme.palette.text.hint` here.

**Review by mazursasha1990 — CHANGES_REQUESTED** [review]
Added more comments.

File: src/react/src/util/styles.js (hunk 19 above)
Comment (Innavin369) [reply src/react/src/util/styles.js:725]: done

File: src/react/src/util/styles.js (hunk 18 above)
Comment (Innavin369) [reply src/react/src/util/styles.js:721]: done

File: src/react/src/util/styles.js (hunk 17 above)
Comment (Innavin369) [reply src/react/src/util/styles.js:719]: done

File: src/react/src/util/styles.js (hunk 20 above)
Comment (Innavin369) [reply src/react/src/util/styles.js:669]: done

**Review by roman-stolar — DISMISSED** [review]
LGTM, please check CodeRabbit comments and resolved if it unnecessary.

**Comment by roman-stolar** [comment]
> ## [Quality Gate Failed](https://sonarcloud.io/dashboard?id=opensupplyhub_open-supply-hub&pullRequest=397) **Quality Gate failed**
> Failed conditions  [8.6% Duplication on New Code](https://sonarcloud.io/component_measures?id=opensupplyhub_open-supply-hub&pullRequest=397&metric=new_duplicated_lines_density&view=list) (required ≤ 3%)
> 
> [See analysis details on SonarCloud](https://sonarcloud.io/dashboard?id=opensupplyhub_open-supply-hub&pullRequest=397)

@Innavin369 Please add `// NOSONAR: justification for ignoring specific code` to the lines of duplicated code to be passed by SonarQube.

**Comment by mazursasha1990** [comment]
@Innavin369 If you intend to add a comment for ignoring specific code, please do so on the mocked data that will be removed after the integration task is implemented. Otherwise, you can modify the data in one of the files (tests or mocks) to prevent repetition.

File: src/react/src/reducers/DashboardContributionRecordReducer.js [hunk 21]
```diff
@@ -0,0 +1,76 @@
...
+    failFetchingModerationEvent,
+    completeFetchingModerationEvent,
+    startFetchingPotentialMatches,
+    failFetchingPotentialMatches,
+    completeFetchingPotentialMatches,
+    cleanupContributionRecord,
+} from '../actions/dashboardContributionRecord';
+
+const initialState = Object.freeze({
+    moderationEvent: Object.freeze({
```
Comment (mazursasha1990) [inline src/react/src/reducers/DashboardContributionRecordReducer.js:15]: Please use `singleModerationEvent`.

File: src/react/src/actions/dashboardContributionRecord.js [hunk 22]
```diff
@@ -0,0 +1,130 @@
...
+            alpha_2: 'DE',
+            alpha_3: 'DEU',
+            numeric: '276',
+        },
+        claim_status: 'unclaimed',
+    },
+];
+
+// eslint-disable-next-line no-unused-vars
+export function fetchModerationEvent(moderationID) {
```
Comment (mazursasha1990) [inline src/react/src/actions/dashboardContributionRecord.js:89]: Please change it to `fetchSingleModerationEvent`.

**Review by mazursasha1990 — CHANGES_REQUESTED** [review]
Please fix the warnings in the browser console:
Screenshot from 2024-11-11 19-32-59
Screenshot from 2024-11-11 19-32-37

File: src/react/src/components/Dashboard/DashboardContributionRecord.jsx [hunk 23]
```diff
@@ -0,0 +1,262 @@
...
+                        <div className={classes.emptyBlockStyles}>
+                            <Typography
+                                className={classes.emptyTextStyle}
+                                variant="title"
+                            >
+                                No potential matches found
+                            </Typography>
+                        </div>
+                    ) : (
+                        <List styles={styles.recordList}>
```
[the remark this answers is not in the document]
Comment (Innavin369) [reply src/react/src/components/Dashboard/DashboardContributionRecord.jsx:102]: Doublechecked. Removed

File: src/react/src/components/Dashboard/DashboardContributionRecord.jsx [hunk 24]
```diff
@@ -0,0 +1,262 @@
...
+    fetchPotentialMatchError,
+}) => {
+    useEffect(() => {
+        fetchEvent();
+        fetchMatches();
+    }, [fetchEvent, fetchMatches]);
+
+    if (fetchEventError) {
+        return (
+            <Typography variant="body2" style={{ color: 'red' }}>
```
[the remark this answers is not in the document]
Comment (Innavin369) [reply src/react/src/components/Dashboard/DashboardContributionRecord.jsx:53]: moved to classes

File: src/react/src/reducers/DashboardContributionRecordReducer.js (hunk 21 above)
Comment (Innavin369) [reply src/react/src/reducers/DashboardContributionRecordReducer.js:15]: renamed

File: src/react/src/actions/dashboardContributionRecord.js (hunk 22 above)
Comment (Innavin369) [reply src/react/src/actions/dashboardContributionRecord.js:89]: renamed

File: src/react/src/components/Dashboard.jsx [hunk 25]
```diff
@@ -255,7 +265,12 @@ function Dashboard({ userWithAccessHasSignedIn, fetchingSessionSignIn }) {
                         exact
                         path={dashboardModerationQueueRoute}
                         component={DashboardModerationQueue}
-                    />
+                    />{' '}
```
[the remark this answers is not in the document]
Comment (Innavin369) [reply src/react/src/components/Dashboard.jsx:268]: removed

File: src/react/src/actions/dashboardContributionRecord.js [hunk 26]
```diff
@@ -0,0 +1,132 @@
...
+
+export const startFetchingModerationEvent = createAction(
+    'START_FETCHING_MODERATION_EVENT',
+);
+export const failFetchingModerationEvent = createAction(
+    'FAIL_FETCHING_MODERATION_EVENT',
+);
+export const completeFetchingModerationEvent = createAction(
+    'COMPLETE_FETCHING_MODERATION_EVENT',
+);
```
[the remark this answers is not in the document]
Comment (Innavin369) [reply src/react/src/actions/dashboardContributionRecord.js:12]: done

**Review by roman-stolar — APPROVED** [review]
LGTM

**Pull request merged**